Return a promise from deactivation to ensure vscode will wait until P…

daserge opened the pull request:
…ackager process is killed

VS Code expects a promise returned from deactivate and waits until it resolved or a timeout occurs before exiting.
Not returning a promise can cause a situation when we have not yet killed Packager while other extensions' deactivates have finished and vscode process exits - this will result in the Packager process being detached from parent and hang in the process tree root.

This closes #300

msftclas commented:
Hi __@daserge__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Sergey Shakhnazarov (Akvelon)). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

TimBarham commented:
Great find, @daserge! I looked at the vscode source and can confirm your findings, that the call to `deactivate()` expects a promise be returned (though I couldn't find it documented anywhere).

My only comment would be that I think it would be preferable to `Q.Promise()` rather than the Q specific `Q.defer`, as in:

```js
export function deactivate(): Q.Promise<any> {
    return Q.Promise(function (resolve) {
        // Kill any packager processes that we spawned
        entryPointHandler.runFunction("extension.deactivate",
            ErrorHelper.getInternalError(InternalErrorCode.FailedToStopPackagerOnExit),
            () => {
                commandPaletteHandler.stopPackager().done(() => {
                    // Tell vscode that we are done with deactivation
                    resolve();
                });
            }, /*errorsAreFatal*/ true);
    });
}
```

Also, I don't believe you need to pass any value with the `resolve()` call.

daserge commented:
Thanks Tim!
I've addressed the comments.
Regarding the `resolve` return value - it is required according to the typings:
https://github.com/Microsoft/vscode-react-native/blob/61c4db14ec615d4eb5ada64fc869385489315d29/src/typings/q/Q.d.ts#L278

TimBarham commented:
Can't you just declare this method as returning `Q.Promise<void>`? There doesn't seem to be any requirement on the vscode side for the promise to resolve to a value.

daserge commented:
@TimBarham - thanks!
Updated to `Q.Promise<void>`.

TimBarham commented on tsconfig.json:
```diff
@@ -17,6 +17,7 @@
         "tools/",
-        "src/test/resources/sampleReactNative022Project/"
+        "src/test/resources/sampleReactNative022Project/",
+        ".vscode-test"
```
Can you clarify what this change was needed for?

daserge replied:
It was needed to exclude test artifacts from tslint checks.

TimBarham replied:
Ah, fair enough.

daserge commented on src/extension/rn-extension.ts:
```diff
@@ -83,13 +83,18 @@ export function activate(context: vscode.ExtensionContext): void {
+            () => {
+                commandPaletteHandler.stopPackager().done(() => {
+                    // Tell vscode that we are done with deactivation
+                    resolve(undefined);
```
Looks like [void 0](https://github.com/Microsoft/vscode-react-native/search?utf8=%E2%9C%93&q=resolve%28void+0%29&type=Code) is used for this case - what do you think?

TimBarham replied:
I saw both were used, but `null` just seemed cleaner 😄

daserge replied:
Updated to `void 0` as it seems to be more commonly used in this repository.

TimBarham replied:
Fair enough.
